clang/test/SemaSYCL/intel-max-global-work-dim-device.cpp

```diff
@@ -47,14 +61,109 @@ struct TRIFuncObjGood2 {
+// In case the value of 'max_global_work_dim' attribute equals to 0 we shall
+// ensure that if max_work_group_size and reqd_work_group_size attributes exist,
+// they hold equal values (1, 1, 1).
+
 struct TRIFuncObjBad {
   [[intel::max_global_work_dim(0)]]
   [[intel::max_work_group_size(8, 8, 8)]] // expected-error{{'max_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
   [[cl::reqd_work_group_size(4, 4, 4)]]   // expected-error{{'reqd_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
   void
   operator()() const {}
 };
+
+struct TRIFuncObjBad1 {
+  [[intel::max_work_group_size(8, 8, 8)]] // expected-error{{'max_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjBad2 {
+  [[cl::reqd_work_group_size(4, 4, 4)]]   // expected-error{{'reqd_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjBad3 {
+  [[intel::reqd_work_group_size(4, 4, 4)]]   // expected-error{{'reqd_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+// Tests for incorrect argument values for Intel FPGA function attributes:
+// reqd_work_group_size, max_work_group_size and max_global_work_dim.
+
+struct TRIFuncObjBad4 {
+  // expected-error@+2{{'reqd_work_group_size' X-, Y- and Z- sizes must be 1 when 'max_global_work_dim' attribute is used with value 0}}
+  // expected-warning@+1{{implicit conversion changes signedness: 'int' to 'unsigned long long'}}
+  [[intel::reqd_work_group_size(4, 4, -4)]]
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjBad5 {
+  [[intel::max_work_group_size(4, 4, 4.f)]] // expected-error{{integral constant expression must have integral or unscoped enumeration type, not 'float'}}
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjBad6 {
+  [[cl::reqd_work_group_size(0, 4, 4)]] // expected-error{{'reqd_work_group_size' attribute must be greater than 0}}
+  [[intel::max_global_work_dim(0)]] void
+  operator()() const {}
+};
+
+struct TRIFuncObjBad7 {
+  [[intel::reqd_work_group_size(4, 4, 4)]]
+  [[intel::max_global_work_dim(-2)]] // expected-error{{'max_global_work_dim' attribute requires a non-negative integral compile time constant expression}}
+  void operator()() const {}
+};
+
+struct TRIFuncObjBad8 {
+  [[intel::max_work_group_size(4, 4, 4)]]
+  [[intel::max_global_work_dim(4.f)]] // expected-error{{integral constant expression must have integral or unscoped enumeration type, not 'float'}}
+  void operator()() const {}
+};
```

Comment:
Test cases caused assertions with existing support of max_global_work_dim attribute. With new design, the test cases generate expected diagnostic.